addons/binding/org.openhab.binding.nibeheatpump/src/main/java/org/openhab/binding/nibeheatpump/internal/message/ModbusDataReadOutMessage.java

```diff
@@ -0,0 +1,143 @@
+ *
+ * @author Pauli Anttila - Initial contribution
+ */
+public class ModbusDataReadOutMessage extends NibeHeatPumpBaseMessage {
+
+    List<ModbusValue> values;
+
+    private ModbusDataReadOutMessage(MessageBuilder builder) {
+        super.msgType = MessageType.MODBUS_DATA_READ_OUT_MSG;
+        this.values = builder.values;
+    }
+
+    public ModbusDataReadOutMessage(byte[] data) throws NibeHeatPumpException {
+        encodeMessage(data);
+    }
+
+    public List<ModbusValue> getValues() {
+        return values;
+    }
+
+    @Override
+    public void encodeMessage(byte[] data) throws NibeHeatPumpException {
+        values = ParseMessage(data);
+    }
+
+    @Override
+    public byte[] decodeMessage() {
+        return createDataReadOutPdu(values);
+    }
+
+    private byte[] createDataReadOutPdu(List<ModbusValue> values) {
+
+        byte datalen = (byte) (values.size() * 4);
+        byte msglen = (byte) (6 + datalen);
+
+        byte[] data = new byte[msglen];
+
+        data[0] = NibeHeatPumpProtocol.FRAME_START_CHAR_FROM_NIBE;
+        data[1] = 0x00;
+        data[2] = NibeHeatPumpProtocol.ADR_MODBUS40;
+        data[3] = NibeHeatPumpProtocol.CMD_MODBUS_DATA_MSG;
+        data[4] = datalen;
+
+        int i = NibeHeatPumpProtocol.OFFSET_DATA;
+
+        for (ModbusValue value : values) {
+
+            int coildAddress = value.getCoilAddress();
+            int val = value.getValue();
+
+            data[i + 0] = (byte) (coildAddress & 0xFF);
+            data[i + 1] = (byte) ((coildAddress >> 8) & 0xFF);
+            data[i + 2] = (byte) (val & 0xFF);
+            data[i + 3] = (byte) ((val >> 8) & 0xFF);
+            i += 4;
+        }
+
+        data[msglen - 1] = NibeHeatPumpProtocol.calculateChecksum(data, 2, msglen);
+
+        return data;
+    }
+
+    @Override
+    public String toString() {
+        String str = "";
+
+        str += super.toString();
+        str += ", Values: ";
+        str += values.toString();
+        return str;
+    }
+
+    private List<ModbusValue> ParseMessage(byte[] data) throws NibeHeatPumpException {
+        if (NibeHeatPumpProtocol.isModbus40DataReadOut(data)) {
+            super.encodeMessage(data);
+            data = super.rawMessage;
+            final int datalen = data[NibeHeatPumpProtocol.OFFSET_LEN];
+            final int msglen = 5 + datalen;
+
+            List<ModbusValue> vals = new ArrayList<ModbusValue>();
+
+            try {
+                for (int i = NibeHeatPumpProtocol.OFFSET_DATA; i < (msglen - 1); i += 4) {
+
+                    int id = ((data[i + 1] & 0xFF) << 8 | (data[i + 0] & 0xFF));
+                    int value = (short) ((data[i + 3] & 0xFF) << 8 | (data[i + 2] & 0xFF));
+
+                    if (id != 0xFFFF) {
+                        vals.add(new ModbusValue(id, value));
+                    }
+                }
+            } catch (ArrayIndexOutOfBoundsException e) {
```

Comment:
I agree, all codes would be better if every one write code which doesn't have bugs ;)